File: src/main/java/com/kropiejohn/ftc/glasses/model/GlassesDatabase.java
package com.kropiejohn.ftc.glasses.model;

import com.kropiejohn.ftc.glasses.FtcGlasses;
import javafx.collections.FXCollections;
import javafx.collections.MapChangeListener;
import javafx.collections.ObservableMap;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Date;
import java.util.List;

public enum GlassesDatabase {
    INSTANCE(FXCollections.observableHashMap());

    final MapChangeListener<Integer, Glasses> listener = change -> save();

    public int getNewIndex() {
        return db.size() + 1;
    }

    ObservableMap<Integer, Glasses> db;

    GlassesDatabase(ObservableMap<Integer, Glasses> db) {
        this.db = db;

        restore();
        db.addListener(listener);
    }

    public ObservableMap<Integer, Glasses> get() {
        return db;
    }

    public void replaceContents(List<Glasses> toAdd) {
        db.removeListener(listener);
        db.clear();
        toAdd.forEach(glasses -> db.put(glasses.getNumber(), glasses));
        save();
        db.addListener(listener);
    }

    private void save() {
        var glassesList = new JSONArray();
        //noinspection unchecked
        db.values().forEach(v -> glassesList.add(getGlassesJsonObj(v)));

        try (var writer = new FileWriter(FtcGlasses.class.getResource("database/GlassesDatabase.json").getFile())) {
            writer.write(glassesList.toJSONString());
            writer.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void restore() {
        var jsonParse = new JSONParser();
        try (var reader = new FileReader(FtcGlasses.class.getResource("database/GlassesDatabase.json").getFile())) {
            var obj = jsonParse.parse(reader);
            var glassesList = (JSONArray) obj;
            for (Object glassesJson : glassesList) {
                var glasses = restoreGlasses((JSONObject) glassesJson);
                db.put(glasses.getNumber(), glasses);
            }
        } catch (IOException | ParseException e) {
            e.printStackTrace();
        }
    }

    private Glasses restoreGlasses(JSONObject glassesJson) {
        var glasses = new Glasses();

        glasses.setRightSphere((Double) glassesJson.get("rightSphere"));
        glasses.setRightCylinder((Double) glassesJson.get("rightCylinder"));
        glasses.setRightAxis(Math.toIntExact((Long) glassesJson.get("rightAxis")));
        glasses.setRightFocal(Math.toIntExact((Long) glassesJson.get("rightFocal")));

        glasses.setLeftSphere((Double) glassesJson.get("leftSphere"));
        glasses.setLeftCylinder((Double) glassesJson.get("leftCylinder"));
        glasses.setLeftAxis(Math.toIntExact((Long) glassesJson.get("leftAxis")));
        glasses.setLeftFocal(Math.toIntExact((Long) glassesJson.get("leftFocal")));

        glasses.setGender(Gender.valueOf((String) glassesJson.get("sex")));
        glasses.setAge(Age.valueOf((String) glassesJson.get("age")));
        glasses.setBifocals(YesNo.valueOf((String) glassesJson.get("bifocals")));

        glasses.setNumber(Math.toIntExact((Long) glassesJson.get("number")));
        glasses.setEntryDate(new Date((Long) glassesJson.get("entryDate")));
        glasses.setRemoved((Boolean) glassesJson.get("removed"));

        return glasses;
    }

    @SuppressWarnings("unchecked")
    private JSONObject getGlassesJsonObj(final Glasses glasses) {
        var glassesJson = new JSONObject();
        glassesJson.put("rightSphere", glasses.getRightSphere());
        glassesJson.put("rightCylinder", glasses.getRightCylinder());
        glassesJson.put("rightAxis", glasses.getRightAxis());
        glassesJson.put("rightFocal", glasses.getRightFocal());

        glassesJson.put("leftSphere", glasses.getLeftSphere());
        glassesJson.put("leftCylinder", glasses.getLeftCylinder());
        glassesJson.put("leftAxis", glasses.getLeftAxis());
        glassesJson.put("leftFocal", glasses.getLeftFocal());

        glassesJson.put("sex", glasses.getGender().name());
        glassesJson.put("age", glasses.getAge().name());
        glassesJson.put("bifocals", glasses.getBifocals().name());

        glassesJson.put("number", glasses.getNumber());
        glassesJson.put("entryDate", glasses.getEntryDate().getTime());
        glassesJson.put("removed", glasses.isRemoved());

        return glassesJson;
    }
}

File: src/main/java/com/kropiejohn/ftc/glasses/model/GlassesDatabase.kt
package com.kropiejohn.ftc.glasses.model

import java.sql.Connection
import java.sql.DriverManager
import java.sql.SQLException
import java.util.*
import kotlin.collections.HashMap

object GlassesDatabase {
    private val db = HashMap<Int, Glasses>()
    internal lateinit var connection: Connection
    const val tableName = "GlassesDB"

    init {
        try {
            Class.forName("org.apache.derby.jdbc.EmbeddedDriver")
            connection = DriverManager.getConnection("jdbc:derby:$tableName;create=true")
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        } catch (e: SQLException) {
            e.printStackTrace()
        }

        restore()
    }

    fun get(id: Int): Glasses? {
        return db.get(id)
    }

    fun createUpdate(glasses: Glasses) {
        db.put(glasses.number, glasses)
        save(listOf(glasses))
    }

    fun getAll(): Collection<Glasses> {
        return db.values
    }

    fun getNewIndex(): Int {
        return db.size + 1
    }

    fun containsKey(key: Int): Boolean {
        return db.containsKey(key)
    }

    fun isEmpty(): Boolean {
        return db.isEmpty()
    }

    private fun restore() {
        try {// Create table to start.
            val sqlCreate = "CREATE TABLE $tableName (ID INTEGER, " +
                    "rightSphere DOUBLE, rightCylinder DOUBLE, rightAxis INTEGER, rightFocal INTEGER, leftSphere DOUBLE, " +
                    "leftCylinder DOUBLE, leftAxis INTEGER, leftFocal INTEGER, entryDate DATE, removed CHAR, " +
                    "sex CHAR, age CHAR)"
            val statement = connection.createStatement()
            statement.executeUpdate(sqlCreate)
        } catch (e: SQLException) {
            // Squash table already exists errors.
        }

        val getGlassesSql = "SELECT * FROM $tableName"
        val statement = connection.createStatement()
        val glassesRs = statement.executeQuery(getGlassesSql)

        while (glassesRs.next()) {
            val newGlasses = Glasses()
            newGlasses.number = glassesRs.getInt("ID")
            newGlasses.rightSphere = glassesRs.getDouble("rightSphere")
            newGlasses.rightCylinder = glassesRs.getDouble("rightCylinder")
            newGlasses.rightAxis = glassesRs.getInt("rightAxis")
            newGlasses.rightFocal = glassesRs.getInt("rightFocal")

            newGlasses.leftSphere = glassesRs.getDouble("leftSphere")
            newGlasses.leftCylinder = glassesRs.getDouble("leftCylinder")
            newGlasses.leftAxis = glassesRs.getInt("leftAxis")
            newGlasses.leftFocal = glassesRs.getInt("leftFocal")

            newGlasses.entryDate = Date(glassesRs.getDate("entryDate").time)
            newGlasses.isRemoved = glassesRs.getString("removed") == YesNo.YES.abbreviation
            newGlasses.gender = Gender.MALE.getEnumFromAbbreviationOrName(glassesRs.getString("sex"))
            newGlasses.age = Age.ADULT.getEnumFromAbbreviationOrName(glassesRs.getString("age"))

            db.put(newGlasses.number, newGlasses)
        }
    }

    private fun save(toSave: Collection<Glasses>) {
        var statement = connection.createStatement()
        val sql = "DELETE FROM $tableName"
        // Execute deletion
        statement.executeUpdate(sql)

        val insertSql = "INSERT INTO $tableName (ID , rightSphere, rightCylinder, rightAxis, rightFocal, " +
                "leftSphere, leftCylinder, leftAxis, leftFocal, entryDate, removed, sex, age) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"
        val insertStatement = connection.prepareStatement(insertSql)
        connection.autoCommit = false
        toSave.forEach {
            var counter = 0
            insertStatement.setInt(++counter, it.number)
            insertStatement.setDouble(++counter, it.getRightSphere())
            insertStatement.setDouble(++counter, it.getRightCylinder())
            insertStatement.setInt(++counter, it.getRightAxis())
            insertStatement.setInt(++counter, it.getRightFocal())
            insertStatement.setDouble(++counter, it.getLeftSphere())
            insertStatement.setDouble(++counter, it.getLeftCylinder())
            insertStatement.setInt(++counter, it.getLeftAxis())
            insertStatement.setInt(++counter, it.getLeftFocal())
            insertStatement.setDate(++counter, java.sql.Date(it.entryDate.time))
            insertStatement.setString(++counter, YesNo.fromBoolean(it.isRemoved).abbreviation)
            insertStatement.setString(++counter, it.gender.abbreviation)
            insertStatement.setString(++counter, it.age.abbreviation)
            insertStatement.addBatch()
        }

        insertStatement.executeBatch()
        connection.commit()
    }

    fun replaceContents(glasses: List<Glasses>) {
        db.clear()
        glasses.forEach { db.put(it.number, it) }
        save(db.values)
    }
}